Added support for Redis Sentinel

neoaggelos opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Implements support for Redis Sentinel. Closes #1635 

#### Changes
<!-- What are the changes made in this pull request? -->

- Added new options `redis.failover.enable`, `redis.failover.addreses`, `redis.failover.master-name`
- When user uses the Redis Sentinel options, then a `redis.Client` is created using `redis.NewFailoverClient()` instead of `redis.NewClient()`.
- Updated `config.Redis.IsZero()`
- Updated relevant documentation

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

- `redis.NewFailoverClient()` returns a `redis.Client`, just like `redis.NewClient()`. This means that The Things Stack can trasparently use either Redis or Redis Sentinel without further changes.
- It may be desired to rename the configuration options related to Redis Sentinel.
- Default value for `redis.enable-failover` is set to false, so that the change is backwards compatible and does not break existing installations.
- Some kind of test should be added.
- Redis Sentinel comes with some log messags printed out by default, mixed with log messages from the stack. This should probably be fixed as well. 
- meta: Is the `all` tag correct?

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [ ] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [X] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

johanstokking commented on pkg/config/shared.go:
```diff
@@ -118,14 +118,20 @@ type HTTP struct {
+	Password           string   `name:"password" description:"Password of the Redis server"`
+	Database           int      `name:"database" description:"Redis database to use"`
+	Namespace          []string `name:"namespace" description:"Namespace for Redis keys"`
+	Sentinel           bool     `name:"sentinel" description:"Connect to Redis Sentinel"`
```
```suggestion
	EnableFailover     bool     `name:"enable-failover" description:"Enable failover using Redis Sentinel"`
```

neoaggelos replied:
I have updated the code as per the suggestion

johanstokking commented on pkg/config/shared.go:
```diff
@@ -118,14 +118,20 @@ type HTTP struct {
+	Database           int      `name:"database" description:"Redis database to use"`
+	Namespace          []string `name:"namespace" description:"Namespace for Redis keys"`
+	Sentinel           bool     `name:"sentinel" description:"Connect to Redis Sentinel"`
+	SentinelAddresses  []string `name:"sentinel-addresses" description:"Redis Sentinel server addresses (when sentinel=true)"`
```
`FailoverAddresses`

johanstokking commented on pkg/config/shared.go:
```diff
@@ -118,14 +118,20 @@ type HTTP struct {
+	Namespace          []string `name:"namespace" description:"Namespace for Redis keys"`
+	Sentinel           bool     `name:"sentinel" description:"Connect to Redis Sentinel"`
+	SentinelAddresses  []string `name:"sentinel-addresses" description:"Redis Sentinel server addresses (when sentinel=true)"`
+	SentinelMasterName string   `name:"sentinel-master-name" description:"Redis Sentinel master name (when sentinel=true)"`
```
`FailoverMasterName`

johanstokking commented on pkg/config/shared.go:
```diff
@@ -118,14 +118,20 @@ type HTTP struct {
 // IsZero returns whether the Redis configuration is empty.
-func (r Redis) IsZero() bool { return r.Address == "" && r.Database == 0 && len(r.Namespace) == 0 }
+func (r Redis) IsZero() bool {
+	return (r.Sentinel == true && r.SentinelMasterName == "" && len(r.SentinelAddresses) == 0) ||
```
Use go idiomatic boolean checks (i.e. no `== true` or `== false`)

johanstokking commented:
> * Some kind of test should be added.

@htdvisser @rvolosatovs should we make this part of CI?

> * Redis Sentinel comes with some log messags printed out by default, mixed with log messages from the stack. This should probably be fixed as well.

I suppose you refer to Docker Compose logs? Or is the Redis client printing logs directly?

neoaggelos commented:
> I suppose you refer to Docker Compose logs? Or is the Redis client printing logs directly?

No, this is the go-redis module logging stuff to stderr directly. Example:

```
 DEBUG Creating listener                        address=:8887
 DEBUG Start serving                            address=[::]:1887
 DEBUG Start serving                            address=[::]:8887
 DEBUG Loaded TLS certificate                  
  INFO Setting up Network Server               
redis: 2019/11/23 11:12:54 sentinel.go:379: sentinel: discovered new sentinel="bf962e747d4c658689a02f7747eff779b4db08e3" for master="mymaster"
redis: 2019/11/23 11:12:54 sentinel.go:379: sentinel: discovered new sentinel="5e7bb518d84c1638223c57f4c1b873ae180f874e" for master="mymaster"
redis: 2019/11/23 11:12:54 sentinel.go:379: sentinel: discovered new sentinel="fd196c1a327a9db12849d19e3f4a0958dc27237b" for master="mymaster"
redis: 2019/11/23 11:12:54 sentinel.go:379: sentinel: discovered new sentinel="fd196c1a327a9db12849d19e3f4a0958dc27237c" for master="mymaster"
redis: 2019/11/23 11:12:54 sentinel.go:332: sentinel: new master="mymaster" addr="10.121.205.250:6379"
  INFO Setting up Application Server           
 DEBUG Creating listener                        address=:1883
 DEBUG Creating listener                        address=:8883
 DEBUG Start serving                            address=[::]:1883
 DEBUG Loaded TLS certificate                  
  INFO Setting up Join Server                  
```

The easy solution is to `2> /dev/null`, but I think extra care need be taken.

The go-redis module has a `redis.SetLogger()` that can be used for this, but it requires a `log.Logger`, so passing the custom TTN logger cannot work. What do you suggest? I think that we could either add on option to silence those logs (e.g. `--redis.disable-failover-logs`), or try to somehow integrate with the TTN logger (this could need changes in `go-redis` itself).

johanstokking commented:
Thanks for clarifying!

> The go-redis module has a `redis.SetLogger()` that can be used for this, but it requires a `log.Logger`, so passing the custom TTN logger cannot work

So we basically need to implement the indirection. That seems to be useful anyway to do in `pkg/log`.

johanstokking commented on cmd/internal/shared/config.go:
```diff
@@ -76,9 +76,12 @@ var DefaultGRPCConfig = config.GRPC{
+	Database:           0,
+	Namespace:          []string{"ttn", "v3"},
+	EnableFailover:     false,
+	FailoverMasterName: "mymaster",
```
I think it's cleaner if we require the user to specify something rather than `mymaster`. Our aim is to keep default settings sane for local use.

neoaggelos replied:
I had added `mymaster` as default because it is the typical default master name for Redis Sentinel. I agree with your suggestion, removed the default value

johanstokking commented on cmd/internal/shared/config.go:
```diff
@@ -76,9 +76,12 @@ var DefaultGRPCConfig = config.GRPC{
+	Namespace:          []string{"ttn", "v3"},
+	EnableFailover:     false,
+	FailoverMasterName: "mymaster",
+	FailoverAddresses:  []string{},
```
nit; `nil` is a bit cleaner

neoaggelos replied:
Removed default empty list value as well

johanstokking commented on doc/content/reference/configuration/the-things-stack.md:
```diff
@@ -105,11 +105,22 @@ The `blob` source loads from the given path in a bucket. This requires the globa
 - `redis.database`: Redis database to use
 - `redis.namespace`: Namespace for Redis keys
 
+Or you can enable failover using Redis Sentinel:
```
nit; add link to Redis Sentinel docs

johanstokking commented on pkg/redis/redis.go:
```diff
@@ -78,15 +78,29 @@ type Config struct {
+func NewRedisClient(conf config.Redis) *redis.Client {
+	if conf.EnableFailover {
+		return redis.NewFailoverClient(&redis.FailoverOptions{
+			MasterName:    conf.FailoverMasterName,
```
But there's no error checking if this is empty, right?

Shouldn't we fail if failover is set to true but a master name is not defined? That is now silently considered `IsZero()`

neoaggelos replied:
If the user enables failover but has invalid (or missing) arguments, then the client creation fails:

```
$ go run ./cmd/ttn-lw-stack start --redis.enable-failover=true  
  INFO Setting up core component               
(.........)        
  INFO Setting up Network Server               
error:cmd/internal/shared:initialize_network_server (could not initialize Network Server)
--- error:pkg/redis:store (store error)
--- redis: all sentinels are unreachable
exit status 255
```

This is consistent with how the stack fails to initiate when Redis is unavailable:

```
$ go run ./cmd/ttn-lw-stack start --redis.address=localhost:16379
  INFO Setting up core component    
(.........) 
  INFO Setting up Network Server               
error:cmd/internal/shared:initialize_network_server (could not initialize Network Server)
--- error:pkg/redis:store (store error)
--- dial tcp 127.0.0.1:16379: connect: connection refused
exit status 255
```

Is this the desired behaviour?

johanstokking replied:
Yes

johanstokking commented:
Please also update release notes

neoaggelos commented:
I have updated the Changelog, and I also disabled the logs from the go-redis module.

htdvisser commented on pkg/config/shared.go:
```diff
@@ -118,14 +118,20 @@ type HTTP struct {
+	Password           string   `name:"password" description:"Password of the Redis server"`
+	Database           int      `name:"database" description:"Redis database to use"`
+	Namespace          []string `name:"namespace" description:"Namespace for Redis keys"`
+	EnableFailover     bool     `name:"enable-failover" description:"Enable failover using Redis Sentinel (default false)"`
```
For consistency with the other options, it would be nice to use

`failover.enable`
`failover.addresses`
`failover.master-name`

in the `name` struct tags.

rvolosatovs replied:
In fact, wrapping the failover config into a separate struct and having it as field on `Redis` is a better approach here IMO.
Also I don't think we specify defaults in the description anywhere else?
```suggestion
	Failover struct{
	   Enable     bool `name:"enable" description:"Enable failover using Redis Sentinel"`
	   Addresses  []string `name:"addresses" description:"Redis Sentinel server addresses"`
	   MasterName string   `name:"master-name" description:"Redis Sentinel master name"`
	} `name:"failover"`
```
(or `type RedisFailover struct`)

neoaggelos replied:
Added options in a new `type RedisFailover struct`. The options are now `--redis.failover.enable`, etc, like @rvolosatovs suggested.

htdvisser commented on pkg/events/redis/redis.go:
```diff
@@ -22,17 +22,15 @@ import (
 	"github.com/go-redis/redis"
 	"go.thethings.network/lorawan-stack/pkg/config"
 	"go.thethings.network/lorawan-stack/pkg/events"
+
```
You can remove this empty line

rvolosatovs commented on pkg/config/shared.go:
```diff
@@ -118,14 +118,20 @@ type HTTP struct {
 // IsZero returns whether the Redis configuration is empty.
-func (r Redis) IsZero() bool { return r.Address == "" && r.Database == 0 && len(r.Namespace) == 0 }
+func (r Redis) IsZero() bool {
+	return (r.EnableFailover && r.FailoverMasterName == "" && len(r.FailoverAddresses) == 0) ||
```
This is not correct - `r.Namespace` and `r.Database` are only checked with `!r.EnableFailover` in proposed implementation, but they have to be checked always

neoaggelos replied:
Yeah, totally missed that. Should be OK now

neoaggelos commented:
Hello,

> Also I don't think we specify defaults in the description anywhere else?

Default values are only shown for truthy-values, so `(default false)` was hidden. Removed it all the same.

> In fact, wrapping the failover config into a separate struct and having it as field on Redis is a better approach here IMO.

Added a `type RedisFailover struct`.

I also rebased against the new changes in `master`.

rvolosatovs commented on pkg/config/shared.go:
```diff
@@ -116,16 +116,28 @@ type HTTP struct {
-func (r Redis) IsZero() bool { return r.Address == "" && r.Database == 0 && len(r.Namespace) == 0 }
+func (r Redis) IsZero() bool {
+	return r.Database == 0 && len(r.Namespace) == 0 &&
+		((r.Failover.Enable && r.Failover.MasterName == "" && len(r.Failover.Addresses) == 0) ||
```
```suggestion
		(r.Failover.Enable && r.Failover.MasterName == "" && len(r.Failover.Addresses) == 0) ||
					(!r.Failover.Enable && r.Address == "")
```

rvolosatovs commented on cmd/internal/shared/config.go:
```diff
@@ -79,6 +79,9 @@ var DefaultRedisConfig = config.Redis{
 	Address:   "localhost:6379",
 	Database:  0,
 	Namespace: []string{"ttn", "v3"},
+	Failover: config.RedisFailover{
```
nit; You can also not set `Failover` for the same result.

htdvisser commented:
Travis is sometimes a bit flaky, no need to worry about that.

@neoaggelos it would be nice if you could do another (interactive) rebase to reword your commit messages so that they are in the imperative mood, so "Fix" instead of "Fixes"/"Fixed" (see also https://github.com/TheThingsNetwork/lorawan-stack/blob/master/CONTRIBUTING.md#commit-messages).

@rvolosatovs will do a final review, and then get this merged.

neoaggelos commented:
Applied changes requested and squashed all changes in a single commit.

rvolosatovs commented:
Applied small diff https://github.com/TheThingsNetwork/lorawan-stack/compare/3e83fb856e5602201fc694bb88ea886dd5068772..8f6155d4dc2435601100000d692242dde2506370 and merging
